benchalerts/benchalerts/alert_pipeline.py

```diff
@@ -0,0 +1,97 @@
+"""Every project implementing ``benchalerts`` may want different input and output
+channels for their alerts. This module implements a relatively simple framework for
+composing alert pipelines out of common steps.
+"""
+
+import abc
+from traceback import format_exc
+from typing import Any, Dict, List, Optional
+
+
+class AlertPipelineStep(abc.ABC):
+    """One step of a benchalerts pipeline.
+
+    Parameters
+    ----------
+    step_name
+        The name for this step. If not given, will default to this class's name.
+    """
+
+    def __init__(self, step_name: Optional[str]) -> None:
+        self.step_name = step_name or self.__class__.__name__
+
+    @abc.abstractmethod
+    def run_step(self, previous_outputs: Dict[str, Any]) -> Any:
+        """Run this step.
+
+        Parameters
+        ----------
+        previous_outputs
+            A dict of previous steps' outputs in the pipeline, keyed by step name.
+
+        Returns
+        -------
+        Any
+            Anything (sent to subsequent steps via the ``previous_outputs`` dict).
```

Comment:
Yes, you're right that this can represent DAGs. This was a change from an earlier idea I had for this PR. Just to clarify, are you suggesting we shouldn't use the term "pipeline" to describe this concept?

I agree people should be careful about complexity, but if they'd like to notify a GitHub PR and a slack channel (with different logic for both) we might need to handle that sort of complexity. But I totally hear your feedback that someone should be able to glance at an instance of one of these and know exactly what's going on.

Reply:
I don't care tremendously what we call it, but I do think it's important with this structure that it's very obvious what dependencies on previous steps are.